src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -3262,6 +3262,293 @@ void C2_MacroAssembler::stringL_indexof_char(Register str1, Register cnt1, Regis
+  emit_int32( -510534177);
+  emit_int32( 1507551809);
+  emit_int32( -505558625);
+  emit_int32( -293403007);
+  emit_int32(  129082719);
+  emit_int32(-1796951359);
+  emit_int32( -196513505);
+  emit_int32(-1807454463);
+  emit_int32( 1742810335);
+  emit_int32(  887503681);
+  emit_int32(   28629151);
+  emit_int32(     923521);
+  emit_int32(      29791);
+  emit_int32(        961);
+  emit_int32(         31);
+  emit_int32(          1);
+
+  bind(LONG);
+
+  // int coef = 1;
+  movl(coef, 1);
+  // int i = cnt1 - 1;
+  movl(index, cnt1);
+  decrementl(index);
+  // bound = cnt1 & ~(32-1);
+  bound = cnt1;
+  movl(bound, cnt1);
+  andl(bound, ~(32-1));
+
+  // result = 0;
+  if (!is_string_hashcode) {
+    movl(result, 0);
+  }
+
+  // for (; i >= bound; i -= 1) {
+  bind(LONG_SCALAR_LOOP_BEGIN);
+  // i >= bound;
+  cmpl(index, bound);
+  jccb(Assembler::less, LONG_SCALAR_LOOP_END);
+  // result += coef * ary1[i];
+  arrays_hashcode_elload(tmp, Address(ary1, index, Address::times(elsize)), eltype, is_string_hashcode);
+  imull(tmp, coef);
+  addl(result, tmp);
+  // coef *= 31;
+  movl(tmp, 31);
+  imull(coef, tmp);
+  // i -= 1;
+  decrementl(index);
+  jmpb(LONG_SCALAR_LOOP_BEGIN);
+  bind(LONG_SCALAR_LOOP_END);
+  // }
+
+  for (int idx = 0; idx < 4; idx++) {
+    // vresult = IntVector.zero(I256);
+    vpxor(vresult[idx], vresult[idx]);
+  }
+  // vnext = IntVector.broadcast(I256, power_of_31_backwards[0]);
+  movdl(vnext, InternalAddress(power_of_31_backwards + (0 * sizeof(jint))));
+  vpbroadcastd(vnext, vnext, Assembler::AVX_256bit);
+  // vcoef = IntVector.fromArray(I256, power_of_31_backwards, 1);
+  for (int idx = 0; idx < 4; idx++) {
+    arrays_hashcode_elvload(vcoef[idx], InternalAddress(power_of_31_backwards + ((8 * idx + 1) * sizeof(jint))), T_INT);
+  }
+  // vcoef *= coef
+  movdl(vtmp0, coef);
+  vpbroadcastd(vtmp0, vtmp0, Assembler::AVX_256bit);
+  for (int idx = 0; idx < 4; idx++) {
+    vpmulld(vcoef[idx], vcoef[idx], vtmp0, Assembler::AVX_256bit);
+  }
+
+  // for (i &= ~(31); i >= 0; i -= 8*4) {
+  // i &= ~(31);
+  andl(index, ~(31));
+  bind(LONG_VECTOR_LOOP_BEGIN);
+  // loop fission to upfront the cost of fetching from memory, OOO execution
+  // can then hopefully do a better job of prefetching
+  for (int idx = 0; idx < 4; idx++) {
+    arrays_hashcode_elvload(vtmp[idx], Address(ary1, index, Address::times(elsize), 8 * idx * elsize), eltype);
+  }
+  // vresult += vcoef * ary1[i+8*idx:i+8*idx+7]; vcoef *= vnext;
+  for (int idx = 0; idx < 4; idx++) {
+    arrays_hashcode_elvcast(vtmp[idx], eltype);
+    vpmulld(vtmp[idx], vtmp[idx], vcoef[idx], Assembler::AVX_256bit);
+    vpaddd(vresult[idx], vresult[idx], vtmp[idx], Assembler::AVX_256bit);
+  }
+  // i -= 32;
+  subl(index, 32);
+  // i >= 0;
+  cmpl(index, 0);
+  jccb(Assembler::less, LONG_VECTOR_LOOP_END);
+  for (int idx = 0; idx < 4; idx++) {
+    vpmulld(vcoef[idx], vcoef[idx], vnext, Assembler::AVX_256bit);
+  }
+  jmp(LONG_VECTOR_LOOP_BEGIN);
```

Comment:
Calculating backward forces you to do calculating the coefficients on each iteration, I think doing this normally would be better.

Reply:
But doing it forward requires a `reduceLane` on each iteration. It's faster to do it backward.

Reply:
Ok, I can try rewriting as @merykitty suggests and compare. I'm running out of time to spend on this right now, though, so I sort of hope we can do this experiment as a follow-up RFE.

Reply:
You’re right, we can go forward indeed.

Reply:
@cl4es i can write the assembly and send it your way if you want

Reply:
@luhenry if you have some time to translate that snippet then feel free!